avro_schema/frontend.lua

```diff
@@ -135,6 +141,54 @@ local copy_schema
 local copy_schema_error
 local copy_schema_location_info
 
+-- add new type definition to scope
+local function scope_add_type(context, typeid, xtype)
+    local scope = context.scope
+    local _, typeid = extract_nullable(typeid)
+    xtype.nullable = nil
+    if context.defined[typeid] == DEFINITION.DEFINED then
+        copy_schema_error('Type name already defined: %s', typeid)
+    end
+    -- make the type always store name in canonical form
```

Comment:
Maybe note that it is need only for 'DEFINED' types, but not 'REFERENCE' ones. It helps self-check for code correctness for future patches.

Reply:
Do not get the point.

Reply:
context.type_state[typeid] = {name = ...}
context.type_stateptype_id .. '*'] = {nullable = ...}

The expectation that the field sets is such is not obvious.

Reply:
Ouch, yep, the original commect was about DEFINED and REFERENCED, that is another thing.

Reply:
The original comment was about

context.type_state[typeid] = {name = ...}
context.type_stateptype_id .. '*'] = {nullable = ...}

for DEFINED and {} for REFERENCED.